src/magic_ter_fur_transform.cpp

```diff
@@ -0,0 +1,287 @@
+
+            while( target_array.has_more() ) {
+                const std::string valid_terrain = target_array.next_string();
+                ter_flag_transform.emplace( valid_terrain, results );
+                if( !message.empty() ) {
+                    ter_flag_message.emplace( valid_terrain, std::make_pair( message, message_good ) );
+                }
+            }
+        }
+    }
+
+    if( jo.has_member( "furniture" ) ) {
+        JsonArray obj_array = jo.get_array( "furniture" );
+        while( obj_array.has_more() ) {
+            JsonObject furn_obj = obj_array.next_object();
+            JsonArray target_array = furn_obj.get_array( "valid_terrain" );
+
+            weighted_int_list<furn_id> results;
+            load_furn_results( furn_obj, "result", results );
+            const std::string message = furn_obj.get_string( "message" );
+            bool message_good = true;
+            if( furn_obj.has_bool( "message_good" ) ) {
+                message_good = furn_obj.get_bool( "message_good" );
+            }
+
+            while( target_array.has_more() ) {
+                const std::string valid_terrain = target_array.next_string();
+                furn_transform.emplace( furn_id( valid_terrain ), results );
+                if( !message.empty() ) {
+                    furn_message.emplace( furn_id( valid_terrain ), std::make_pair( message, message_good ) );
+                }
+            }
+
+            target_array = furn_obj.get_array( "valid_flags" );
+
+            while( target_array.has_more() ) {
+                const std::string valid_terrain = target_array.next_string();
+                furn_flag_transform.emplace( valid_terrain, results );
+                if( !message.empty() ) {
+                    furn_flag_message.emplace( valid_terrain, std::make_pair( message, message_good ) );
+                }
+            }
+        }
+    }
+}
+
+ter_id ter_furn_transform::next_ter( const ter_id ter ) const
+{
+    const std::map<ter_id, weighted_int_list<ter_id>>::const_iterator result_iter =
+                ter_transform.find( ter );
+    if( result_iter == ter_transform.cend() ) {
+        return t_null;
+    }
+    const ter_id *result = result_iter->second.pick();
+    if( result == NULL ) {
+        return t_null;
+    }
+    return *result;
+}
+
+ter_id ter_furn_transform::next_ter( const std::string &flag ) const
+{
+    const std::map<std::string, weighted_int_list<ter_id>>::const_iterator result_iter =
+                ter_flag_transform.find( flag );
+    if( result_iter == ter_flag_transform.cend() ) {
+        return t_null;
+    }
+    const ter_id *result = result_iter->second.pick();
+    if( result == NULL )         {
+        return t_null;
+    }
+    return *result;
+}
+
+cata::optional<furn_id> ter_furn_transform::next_furn( const furn_id ter ) const
+{
+    const std::map<furn_id, weighted_int_list<furn_id>>::const_iterator result_iter =
```

Comment:
And here.

Lots of code repetition here.

Reply:
yeah, i realized it was being duplicated a bunch as i was writing it, but didn't see an obvious way to reduce that.